File: RecyclerViewTemplates/app/src/main/java/hello/world/recyclerviewtemplates/MainActivity.java
package hello.world.recyclerviewtemplates;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private ArrayList<Dictionary> mArrayList;
    private CustomAdapter mAdapter;
    private int cnt = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView mRecyclerView = (RecyclerView) findViewById(R.id.recyclerview_main_list);
        LinearLayoutManager mLinearLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setLayoutManager(mLinearLayoutManager);
        // Set LinearLayout horizontal
        //mRecyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));

        mArrayList = new ArrayList<>();
        mAdapter = new CustomAdapter(mArrayList);
        mRecyclerView.setAdapter(mAdapter);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(mRecyclerView.getContext(),
                mLinearLayoutManager.getOrientation());
        mRecyclerView.addItemDecoration(dividerItemDecoration);

        Button btnInsert = (Button)findViewById(R.id.button_main_insert);
        btnInsert.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
//                for(int i=0; i<10; i++) {
//                    cnt++;
//                    Dictionary data = new Dictionary(cnt + "", "Apple" + cnt, "사과" + cnt);
//                    mArrayList.add(data);
//                    mAdapter.notifyDataSetChanged();
//                }
                AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
                View view = LayoutInflater.from(MainActivity.this)
                        .inflate(R.layout.edit_box, null, false);
                builder.setView(view);

                final Button ButtonSubmit = (Button) view.findViewById(R.id.button_dialog_submit);
                final EditText editTextID = (EditText) view.findViewById(R.id.edittext_dialog_id);
                final EditText editTextEnglish = (EditText) view.findViewById(R.id.edittext_dialog_endlish);
                final EditText editTextKorean = (EditText) view.findViewById(R.id.edittext_dialog_korean);

                ButtonSubmit.setText("삽입");
                final AlertDialog dialog = builder.create();

                ButtonSubmit.setOnClickListener(new View.OnClickListener() {
                    public void onClick(View v) {
                        String strID = editTextID.getText().toString();
                        String strEnglish = editTextEnglish.getText().toString();
                        String strKorean = editTextKorean.getText().toString();

                        Dictionary dict = new Dictionary(strID, strEnglish, strKorean );
                        mArrayList.add(0, dict); //첫번째 줄에 삽입됨
                        //mArrayList.add(dict); //마지막 줄에 삽입됨
                        // 어댑터에서 RecyclerView에 반영하도록 합니다.
                        // notifyItemInserted(): 특정 position에 삽입하고 싶을 때 사용
                        mAdapter.notifyItemInserted(0);
                        //mAdapter.notifyDataSetChanged();
                        dialog.dismiss();
                    }
                });
                dialog.show();
            }
        });
    }
}

File: MVVM_study_1/app/src/main/java/hello/world/mvvm_study/view/MainActivity.kt
package hello.world.mvvm_study.view

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import hello.world.mvvm_study.R
import hello.world.mvvm_study.adapter.UserListAdapter
import hello.world.mvvm_study.databinding.ActivityMainBinding
import hello.world.mvvm_study.viewmodel.MainViewModel

class MainActivity : AppCompatActivity() {
    private lateinit var viewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_main)        // binding을 통해 layout을 설정하기 때문에 필요없음

        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        val binding = DataBindingUtil.setContentView<ActivityMainBinding>(      // ActivityMainBinding: activity_main.xml의 이름이 자동으로 바뀜
            this,
            R.layout.activity_main
        )
        binding.viewModel = viewModel       // viewModel binding

        // adapter 연결
        val mAdapter = UserListAdapter(this)
        binding.recyclerView.apply {
            adapter = mAdapter
            layoutManager = LinearLayoutManager(applicationContext)
        }

        // viewModel의 LiveData를 관찰하여 UI 업데이트
        viewModel.allUsers.observe(this, Observer { users ->
            users?.let {
                mAdapter.setUsers(it)
            }
        })

        binding.btnAdd.setOnClickListener {
            val dlg = UserDialog(this)
            dlg.show()
        }

        binding.btnRemove.setOnClickListener {
            viewModel.deleteAll()
            mAdapter.notifyDataSetChanged()
        }
    }
}